### PR TITLE
[CPU] Support int8 compute mode in CPU AWQ

### PR DESCRIPTION
**Pull request by yintong-lu**
This PR aims to solve https://github.com/vllm-project/vllm/issues/33797

Improve the performance of AWQ quantized models on CPU backend by replacing the existing INT4 mixed-precision GEMM with higher-performance compute paths.

AWQ - SGLang INT4 W4A8 kernel (adapted from [sglang#8225](https://github.com/sgl-project/sglang/pull/8225) / [sglang#8226](https://github.com/sgl-project/sglang/pull/8226))

- At model-loading time, INT4 weights are repacked into a CPU-optimized layout with pre-computed scale and zero-point via convert_weight_packed_scale_zp.
- During inference, activations are dynamically quantized into uint8 and GEMM is dispatched through a custom AVX512/AMX int4 x int8 kernel (int4_scaled_mm_cpu), which performs dequantization fused inside the GEMM loop.
- Uses environment variable **VLLM_CPU_INT4_W4A8=1** to enable this path.

Cmd to run uts:
python -m pytest tests/kernels/test_awq_int4_to_int8.py -v -s 

Accuracy and performance tests are performed to validate this PR; some results are listed below.
On AWQ path, random dataset is used to verify performance enhancement.
With **input_len=128** and **output_len=128**, this PR records an average throughput of ~**930** tokens/s compared to main branch result of ~**570** tokens/s. 
With **input_len=128** and **output_len=2048**, this PR records an average throughput of ~**525** tokens/s compared to main branch result of ~**290** tokens/s.

### CODE REVIEW

**Comment by yintong-lu** [comment]
This PR is still aiming at functionality so those debug prints would be removed once model-level validation is done

File: vllm/model_executor/layers/quantization/cpu_wna16.py [hunk 1]
```diff
@@ -224,71 +221,147 @@ def create_weights(
...
+        # lyt_debug_A1 no more re-pack to int4; do dequant -> re-quantize int8 instead
         packed_weight = layer.qweight.data
         packed_zeros = layer.qzeros.data
         group_num = packed_zeros.size(0)
         bits = self.quant_config.weight_bits
         pack_factor = int(self.quant_config.pack_factor)
         input_size, packed_output_size = packed_weight.size()
         output_size = packed_output_size * pack_factor
-        isa_hint = _get_isa_hint(layer.scales.dtype)
-        layer.isa_hint = isa_hint
```
Comment (bigPYJ1151) [inline vllm/model_executor/layers/quantization/cpu_wna16.py:242]: I suggest to keep the orignial WOQ path. Perhaps we can spilt the weight process procedures of WOQ and INT8 to different functions for clarity and better maintaince.

File: vllm/model_executor/layers/quantization/cpu_wna16.py [hunk 2]
```diff
@@ -224,71 +221,147 @@ def create_weights(
...
+        n = layer.dnnl_handler.n
+        out = torch.empty((m, n), dtype=x.dtype)
+
+        ops.onednn_scaled_mm(
+            layer.dnnl_handler,
+            x_q,
+            out,
+            x_s,
+            None,           # input_zp (symmetric → no zero point)
+            layer.azp_adj,  # AZP adjustment
```
Comment (bigPYJ1151) [inline vllm/model_executor/layers/quantization/cpu_wna16.py]: ```azp_adj``` is not required because the input is symmetric quantized.

File: vllm/model_executor/kernels/linear/mixed_precision/cpu.py [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+        layer.qweight.data = weight
+
+    # lyt_debug_G1 new method: GPTQ int4 → float → int8 → oneDNN handler
+    def _process_gptq_weights_int8(self, layer: torch.nn.Module):
+        """G1: Convert GPTQ int4 weights to int8 and create onednn handler.
+
+        Flow: unpack int4 → dequant to float (G2) → re-quantize to int8 (G3)
+              → create oneDNN handler (G4)
+        """
+        from vllm.model_executor.layers.quantization.cpu_wna16 import _requantize_to_int8
```
Comment (bigPYJ1151) [inline vllm/model_executor/kernels/linear/mixed_precision/cpu.py]: Feel free to move it to the global scope

File: vllm/model_executor/kernels/linear/mixed_precision/cpu.py [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+        print(f'lyt_debug_G3 final weight_int8 shape: {weight_int8.shape}, '
+            f'dtype: {weight_int8.dtype}')
+        print(f'lyt_debug_G3 final channel_scale shape: {channel_scale.shape}, '
+                f'dtype: {channel_scale.dtype}')
+
+        # lyt_debug_G4 create oneDNN handler for int8 GEMM
+        channel_scale_2d = channel_scale.unsqueeze(0)  # [1, N]
+
+        # AZP adjustment: compensation term for dynamic quantization
+        azp_adj = weight_int8.sum(dim=0, keepdim=True, dtype=torch.float32)
```
Comment (bigPYJ1151) [inline vllm/model_executor/kernels/linear/mixed_precision/cpu.py]: It's not required because of symmetric input quantization

File: vllm/model_executor/kernels/linear/mixed_precision/cpu.py [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+        if self.w_s_name and hasattr(layer, self.w_s_name):
+            setattr(layer, self.w_s_name, None)
+        if self.w_zp_name and hasattr(layer, self.w_zp_name):
+            setattr(layer, self.w_zp_name, None)
+
+        print(f'lyt_debug_G4 _process_gptq_weights_int8 DONE. '
+            f'int8 oneDNN path ready, old int4 params cleaned up.')
+
+    # lyt_debug_g4 flag: use int8 path (True) or fallback to original int4 path (False)
+    _use_int8_path = True
```
Comment (bigPYJ1151) [inline vllm/model_executor/kernels/linear/mixed_precision/cpu.py]: We can add a environment variable for this. I suggest to name it as ```VLLM_CPU_WOQ_INT8_MODE``` (bool). You can refer to ```VLLM_CPU_SGL_KERNEL``` for adding a new env in vLLM.

File: vllm/model_executor/kernels/linear/mixed_precision/cpu.py [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+        m = x_2d.size(0)
+        n = self.dnnl_handler.n
+        out = torch.empty((m, n), dtype=x.dtype)
+        ops.onednn_scaled_mm(
+            self.dnnl_handler,
+            x_q,
+            out,
+            x_s,
+            None,           # input_zp (symmetric → no zero point)
+            self.azp_adj,   #  AZP adjustment
```
Comment (bigPYJ1151) [inline vllm/model_executor/kernels/linear/mixed_precision/cpu.py]: Same as above

File: tests/kernels/test_awq_int4_to_int8.py [hunk 7]
```diff
@@ -0,0 +1,617 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (bigPYJ1151) [inline tests/kernels/test_awq_int4_to_int8.py:1]: We can enable the tests in https://github.com/vllm-project/vllm/blob/main/.buildkite/hardware_tests/cpu.yaml

File: tests/kernels/test_awq_int4_to_int8.py [hunk 8]
```diff
@@ -0,0 +1,617 @@
...
+    _dequant_awq_to_float,
+    _requantize_to_int8,
+)
+from vllm.model_executor.layers.quantization.utils.quant_utils import (
+    pack_cols,
+    unpack_cols,
+)
+
+# oneDNN ops: may not be available outside Docker / CPU build
+_has_onednn = False
```
Comment (bigPYJ1151) [inline tests/kernels/test_awq_int4_to_int8.py]: We can reuse ```_supports_onednn``` in ```_custom_ops .py```

File: .buildkite/hardware_tests/cpu.yaml
Comment (bigPYJ1151) [inline .buildkite/hardware_tests/cpu.yaml]: We can also add a new wna16 e2e smoke test with the INT8 mode

File: vllm/_custom_ops.py [hunk 9]
```diff
@@ -3059,6 +3059,38 @@ def int8_scaled_mm_with_quant_fake(
...
+        w: torch.Tensor,
+        w_zeros: torch.Tensor,
+        w_scales: torch.Tensor,
+        bias: torch.Tensor | None,
+    ) -> torch.Tensor:
+        N = w_scales.size(0) * w_scales.size(-1)
+        return torch.empty((x.size(0), N), dtype=x.dtype, device=x.device)
+
+
+_supports_int4_w4a8 = bool(hasattr(torch.ops._C, "convert_weight_packed_scale_zp"))
```
Comment (bigPYJ1151) [inline vllm/_custom_ops.py]: rename to ```_supports_cpu_w4a8_int8``` for distinguish.

File: vllm/envs.py [hunk 10]
```diff
@@ -52,6 +52,8 @@
     VLLM_CPU_NUM_OF_RESERVED_CPU: int | None = None
     VLLM_CPU_SGL_KERNEL: bool = False
     VLLM_ZENTORCH_WEIGHT_PREPACK: bool = True
+    VLLM_CPU_WOQ_INT8_MODE: bool = False
```
Comment (bigPYJ1151) [inline vllm/envs.py]: ```VLLM_CPU_WOQ_INT8_MODE``` can be removed as numeric errors are too large with direct requantization.

File: vllm/envs.py [hunk 11]
```diff
@@ -52,6 +52,8 @@
...
+    VLLM_CPU_INT4_W4A8: bool = False
```
Comment (bigPYJ1151) [inline vllm/envs.py]: By default it can be set as ```True``` to recover the same behaviour with IPEX kernel.

File: vllm/envs.py [hunk 12]
```diff
@@ -723,6 +725,12 @@ def _get_or_set_default() -> str:
     "VLLM_ZENTORCH_WEIGHT_PREPACK": lambda: bool(
         int(os.getenv("VLLM_ZENTORCH_WEIGHT_PREPACK", "1"))
     ),
+    # (CPU backend only) whether to use int8 oneDNN GEMM for WOQ (AWQ/GPTQ).
+    "VLLM_CPU_WOQ_INT8_MODE": lambda: bool(
+        int(os.getenv("VLLM_CPU_WOQ_INT8_MODE", "0"))
+    ),
```
Comment (bigPYJ1151) [inline vllm/envs.py]: Remove it

File: vllm/envs.py [hunk 13]
```diff
@@ -723,6 +725,12 @@ def _get_or_set_default() -> str:
...
+    # (CPU backend only) whether to use SGLang INT4 W4A8 kernels for AWQ.
+    "VLLM_CPU_INT4_W4A8": lambda: bool(int(os.getenv("VLLM_CPU_INT4_W4A8", "0"))),
```
Comment (bigPYJ1151) [inline vllm/envs.py]: change the default to ```True```

File: vllm/model_executor/kernels/linear/mixed_precision/cpu.py [hunk 14]
```diff
@@ -3,6 +3,7 @@
 
 import torch
 
+import vllm.envs as envs
```
Comment (bigPYJ1151) [inline vllm/model_executor/kernels/linear/mixed_precision/cpu.py]: the changes in this file can be reverted.

File: .buildkite/hardware_tests/cpu.yaml [hunk 15]
```diff
@@ -14,12 +14,16 @@ steps:
   - tests/kernels/attention/test_cpu_attn.py
   - tests/kernels/moe/test_cpu_fused_moe.py
   - tests/kernels/test_onednn.py
+  - tests/kernels/test_awq_int4_to_int8.py
+  - tests/kernels/test_gptq_int4_to_int8.py
```
Comment (bigPYJ1151) [inline .buildkite/hardware_tests/cpu.yaml]: Looks like the kernel tests are not required as SGL tests will cover them. We can extend test cases in ```test_cpu_wna16.py``` to test W4A16 and W4A8 code path.

**Review by bigPYJ1151 — COMMENTED** [review]
Thanks! Overall looks good, just some nits.

File: vllm/model_executor/layers/quantization/cpu_wna16.py [hunk 16]
```diff
@@ -224,7 +223,13 @@ def create_weights(
         layer.register_parameter("scales", scales)
 
     def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
-        torch.set_printoptions(profile="full", linewidth=5000, sci_mode=False)
+        if envs.VLLM_CPU_INT4_W4A8:
```
Comment (bigPYJ1151) [inline vllm/model_executor/layers/quantization/cpu_wna16.py]: Also need to check AMX support as SGL kernels require it. Use ```torch.cpu._is_amx_tile_supported()```.

File: tests/kernels/test_awq_int4_to_int8.py [hunk 17]
```diff
@@ -0,0 +1,277 @@
...
+
+import numpy as np
+import pytest
+import torch
+
+from vllm._custom_ops import _supports_cpu_w4a8_int8
+from vllm.model_executor.layers.quantization.utils.quant_utils import (
+    pack_cols,
+)
+
```
Comment (bigPYJ1151) [inline tests/kernels/test_awq_int4_to_int8.py:31]: Please skip the tests on other backends, ref: https://github.com/vllm-project/vllm/blob/main/tests/kernels/test_onednn.py#L11

Comment (bigPYJ1151) [reply tests/kernels/test_awq_int4_to_int8.py:31]: And enable the tests in the ```CPU-Kernel Tests``` group, https://github.com/vllm-project/vllm/blob/main/.buildkite/hardware_tests/cpu.yaml#L21

**Review by bigPYJ1151 — APPROVED** [review]

File: vllm/model_executor/layers/quantization/cpu_wna16.py [hunk 18]
```diff
@@ -278,13 +281,40 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
...
+        layer.scales = None
+
     def apply(
         self,
         layer: torch.nn.Module,
         x: torch.Tensor,
         bias: torch.Tensor | None = None,
     ) -> torch.Tensor:
-        x = cpu_gemm_wna16(
+        if envs.VLLM_CPU_INT4_W4A8 and torch.cpu._is_amx_tile_supported():
```
Comment (bigPYJ1151) [inline vllm/model_executor/layers/quantization/cpu_wna16.py]: https://buildkite.com/vllm/ci/builds/58919#019d41e4-6575-4a4c-96f0-97237362f081
Looks like we should set ```layer.use_w4a8 = envs.VLLM_CPU_INT4_W4A8 and torch.cpu._is_amx_tile_supported()``` in ```process_weights_after_loading``` to make torch compile work
